homeassistant/components/husqvarna_automower/lawn_mower.py

```diff
@@ -162,3 +193,86 @@ async def async_override_schedule_work_area(
         await self.coordinator.api.commands.start_in_workarea(
             self.mower_id, work_area_id, duration
         )
+
+    @handle_sending_exception()
+    async def async_set_schedule(
+        self,
+        mode: str,
+        start: time,
+        end: time,
+        monday: bool,
+        tuesday: bool,
+        wednesday: bool,
+        thursday: bool,
+        friday: bool,
+        saturday: bool,
+        sunday: bool,
+        work_area_id: int | None = None,
+    ) -> None:
+        """Set schedule."""
+        if start >= end:
+            raise ServiceValidationError(
+                translation_domain=DOMAIN,
+                translation_key="start_after_end",
+            )
+        duration = timedelta(
+            hours=int(end.hour) - int(start.hour),
+            minutes=int(end.minute) - int(start.minute),
+        )
+        user_input = Calendar(
+            start,
+            duration,
+            monday,
+            tuesday,
+            wednesday,
+            thursday,
+            friday,
+            saturday,
+            sunday,
+            work_area_id,
+        )
+        new_list = []
+        existing_data = copy.copy(self.coordinator.data[self.mower_id].calendar.tasks)
+        _LOGGER.debug("existing_data : %s", existing_data)
+        _LOGGER.debug("user_input: %s", user_input)
+        if mode == "overwrite":
+            new_list.append(user_input)
+            tasks = Tasks(tasks=new_list)
+            await self.coordinator.api.commands.set_calendar(self.mower_id, tasks)
+        if mode == "add":
+            for calendar_entry in existing_data:
+                if calendar_entry == user_input:
+                    raise ServiceValidationError(
+                        translation_domain=DOMAIN,
+                        translation_key="calendar_entry_already_exists",
+                    )
+                if work_area_id is not None:
+                    if calendar_entry.work_area_id == user_input.work_area_id:
+                        new_list.append(calendar_entry)
+                if work_area_id is None:
+                    new_list.append(calendar_entry)
```

Comment:
iterating over all existing schedule is a implementation details and should therefore be in the library